File: Common/CecilHelper.vb
Namespace Common

    Public Class CecilHelper

        Public Function IsBaseTypeOf(ByVal Base As Cecil.TypeDefinition, Derived As Cecil.TypeDefinition) As Boolean
            If Base.IsInterface Then
                For Each I In Derived.Interfaces
                    Dim ID = I.Resolve()
                    If Me.Equals(Base, ID) Then
                        Return True
                    End If
                Next
            Else
                Do Until Derived Is Nothing
                    If Me.Equals(Base, Derived) Then
                        Return True
                    End If
                    Base = Derived.BaseType?.Resolve()
                Loop
            End If
            Return False
        End Function

        Public Overloads Function Equals(ByVal Type1 As Cecil.TypeDefinition, ByVal Type2 As Cecil.TypeDefinition) As Boolean
            Return Type1.FullName = Type2.FullName And Type1.Module.FullyQualifiedName = Type2.Module.FullyQualifiedName
        End Function

        Public Function Convert(ByVal Method As Cecil.MethodDefinition) As Reflection.MethodInfo
            Dim Type = Me.Convert(Method.DeclaringType)
            Return Type.GetMethod(Method.Name, Method.Parameters.Select(Function(P) Me.Convert(P.ParameterType.Resolve())).ToArray())
        End Function

        Public Function Convert(ByVal Type As Cecil.TypeDefinition) As Type
            Dim Assembly = Me.Convert(Type.Module.Assembly)
            Return Assembly.GetType(Type.FullName)
        End Function

        Public Function Convert(ByVal Assembly As Cecil.AssemblyDefinition) As Reflection.Assembly
            Return Reflection.Assembly.Load(New Reflection.AssemblyName(Assembly.FullName))
        End Function

        Public Function Convert(ByVal AssemblyName As Cecil.AssemblyNameReference) As Reflection.Assembly
            Return Reflection.Assembly.Load(New Reflection.AssemblyName(AssemblyName.FullName))
        End Function

        Public Function Convert(ByVal Type As Type) As Cecil.TypeDefinition
            Dim FName = Type.FullName
            For Each M In Me.Convert(Type.Assembly).Modules
                Dim T = M.GetType(FName)
                If T IsNot Nothing Then
                    Return T
                End If
            Next
            Throw New Exception()
        End Function

        Public Function Convert(ByVal Assembly As Reflection.Assembly) As Cecil.AssemblyDefinition
            Return Cecil.AssemblyDefinition.ReadAssembly(Assembly.Location)
        End Function

        Public Function GetReferencedAssemblyNames(ByVal Assembly As Cecil.AssemblyDefinition) As Cecil.AssemblyNameReference()
            Return Me.GetRawReferencedAssemblyNames(Assembly).ToArray()
        End Function

        Friend Function GetRawReferencedAssemblyNames(ByVal Assembly As Cecil.AssemblyDefinition) As IEnumerable(Of Cecil.AssemblyNameReference)
            Me.AssemblyNames.Clear()

            For Each M In Assembly.Modules
                For Each A In M.AssemblyReferences
                    Me.AssemblyNames.Add(A)
                Next
            Next

            Return Me.AssemblyNames
        End Function

        Public Function GetReferencedAssemblies(ByVal Assembly As Cecil.AssemblyDefinition) As Cecil.AssemblyDefinition()
            Return Me.GetRawReferencedAssemblies(Assembly).ToArray()
        End Function

        Friend Function GetRawReferencedAssemblies(ByVal Assembly As Cecil.AssemblyDefinition) As IEnumerable(Of Cecil.AssemblyDefinition)
            Me.Assemblies.Clear()

            For Each M In Assembly.Modules
                For Each A In M.AssemblyReferences
                    Me.Assemblies.Add(Me.Resolver.Resolve(A))
                Next
            Next

            Return Me.Assemblies
        End Function

        Public Function GetRecursiveReferencedAssemblies(ByVal Assembly As Cecil.AssemblyDefinition) As Cecil.AssemblyDefinition()
            Return Me.GetRawRecursiveReferencedAssemblies(Assembly).ToArray()
        End Function

        Friend Function GetRawRecursiveReferencedAssemblies(ByVal Assembly As Cecil.AssemblyDefinition) As IEnumerable(Of Cecil.AssemblyDefinition)
            Me.Assemblies.Clear()
            Me.CollectReferences(Assembly)
            Return Me.Assemblies
        End Function

        Private Sub CollectReferences(ByVal Assembly As Cecil.AssemblyDefinition)
            If Not Me.Assemblies.Add(Assembly) Then
                Exit Sub
            End If
            For Each M In Assembly.Modules
                For Each A In M.AssemblyReferences
                    Me.CollectReferences(Me.Resolver.Resolve(A))
                Next
            Next
        End Sub

#Region "Instance Property"
        Private Shared ReadOnly _Instance As CecilHelper = New CecilHelper()

        Public Shared ReadOnly Property Instance As CecilHelper
            Get
                Return _Instance
            End Get
        End Property
#End Region

        Private ReadOnly Resolver As Cecil.DefaultAssemblyResolver = New Cecil.DefaultAssemblyResolver()
        Private ReadOnly Assemblies As HashSet(Of Cecil.AssemblyDefinition) = New HashSet(Of Cecil.AssemblyDefinition)()
        Private ReadOnly AssemblyNames As HashSet(Of Cecil.AssemblyNameReference) = New HashSet(Of Cecil.AssemblyNameReference)()

    End Class

End Namespace

File: Cecil/CecilHelper.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;

namespace Ks.Common
{
    public class CecilHelper
    {
        public bool IsBaseTypeOf(Mono.Cecil.TypeDefinition Base, Mono.Cecil.TypeDefinition Derived)
        {
            if (Base.IsInterface)
            {
                foreach (var I in Derived.Interfaces)
                {
                    var ID = I.InterfaceType.Resolve();
                    if (this.Equals(Base, ID))
                    {
                        return true;
                    }
                }
            }
            else
            {
                while (Derived != null)
                {
                    if (this.Equals(Base, Derived))
                    {
                        return true;
                    }

                    Base = Derived.BaseType?.Resolve();
                }
            }

            return false;
        }

        public bool Equals(Mono.Cecil.TypeDefinition Type1, Mono.Cecil.TypeDefinition Type2)
        {
            return (Type1.FullName == Type2.FullName) & (Type1.Module.FileName == Type2.Module.FileName);
        }

        public System.Reflection.MethodInfo Convert(Mono.Cecil.MethodDefinition Method)
        {
            var Type = this.Convert(Method.DeclaringType);
            return Type.GetMethod(Method.Name, Method.Parameters.Select(P => this.Convert(P.ParameterType.Resolve())).ToArray());
        }

        public Type Convert(Mono.Cecil.TypeDefinition Type)
        {
            var Assembly = this.Convert(Type.Module.Assembly);
            return Assembly.GetType(Type.FullName);
        }

        public System.Reflection.Assembly Convert(Mono.Cecil.AssemblyDefinition Assembly)
        {
            return System.Reflection.Assembly.Load(new System.Reflection.AssemblyName(Assembly.FullName));
        }

        public System.Reflection.Assembly Convert(Mono.Cecil.AssemblyNameReference AssemblyName)
        {
            return System.Reflection.Assembly.Load(new System.Reflection.AssemblyName(AssemblyName.FullName));
        }

        public Mono.Cecil.TypeDefinition Convert(Type Type)
        {
            var FName = Type.FullName;
            foreach (var M in this.Convert(Type.Assembly).Modules)
            {
                var T = M.GetType(FName);
                if (T != null)
                {
                    return T;
                }
            }
            throw new Exception();
        }

        public Mono.Cecil.AssemblyDefinition Convert(System.Reflection.Assembly Assembly)
        {
            return Mono.Cecil.AssemblyDefinition.ReadAssembly(Assembly.Location);
        }

        public Mono.Cecil.AssemblyNameReference[] GetReferencedAssemblyNames(Mono.Cecil.AssemblyDefinition Assembly)
        {
            return this.GetRawReferencedAssemblyNames(Assembly).ToArray();
        }

        internal IEnumerable<Mono.Cecil.AssemblyNameReference> GetRawReferencedAssemblyNames(Mono.Cecil.AssemblyDefinition Assembly)
        {
            this.AssemblyNames.Clear();

            foreach (var M in Assembly.Modules)
            {
                foreach (var A in M.AssemblyReferences)
                {
                    this.AssemblyNames.Add(A);
                }
            }

            return this.AssemblyNames;
        }

        public Mono.Cecil.AssemblyDefinition[] GetReferencedAssemblies(Mono.Cecil.AssemblyDefinition Assembly)
        {
            return this.GetRawReferencedAssemblies(Assembly).ToArray();
        }

        internal IEnumerable<Mono.Cecil.AssemblyDefinition> GetRawReferencedAssemblies(Mono.Cecil.AssemblyDefinition Assembly)
        {
            this.Assemblies.Clear();

            foreach (var M in Assembly.Modules)
            {
                foreach (var A in M.AssemblyReferences)
                {
                    this.Assemblies.Add(this.Resolver.Resolve(A));
                }
            }

            return this.Assemblies;
        }

        public Mono.Cecil.AssemblyDefinition[] GetRecursiveReferencedAssemblies(Mono.Cecil.AssemblyDefinition Assembly)
        {
            return this.GetRawRecursiveReferencedAssemblies(Assembly).ToArray();
        }

        internal IEnumerable<Mono.Cecil.AssemblyDefinition> GetRawRecursiveReferencedAssemblies(Mono.Cecil.AssemblyDefinition Assembly)
        {
            this.Assemblies.Clear();
            this.CollectReferences(Assembly);
            return this.Assemblies;
        }

        private void CollectReferences(Mono.Cecil.AssemblyDefinition Assembly)
        {
            if (!this.Assemblies.Add(Assembly))
            {
                return;
            }

            foreach (var M in Assembly.Modules)
            {
                foreach (var A in M.AssemblyReferences)
                {
                    this.CollectReferences(this.Resolver.Resolve(A));
                }
            }
        }

        public static CecilHelper Instance { get; } = new CecilHelper();

        private readonly Mono.Cecil.DefaultAssemblyResolver Resolver = new Mono.Cecil.DefaultAssemblyResolver();
        private readonly HashSet<Mono.Cecil.AssemblyDefinition> Assemblies = new HashSet<Mono.Cecil.AssemblyDefinition>();
        private readonly HashSet<Mono.Cecil.AssemblyNameReference> AssemblyNames = new HashSet<Mono.Cecil.AssemblyNameReference>();
    }
}
